File: public/engine/dom-contract.ts
// The engine's half of the DOM contract.
//
// The app builds the hosts; the engine paints into them. The two sides never
// call each other — they meet in the DOM, through attribute names, the values
// those attributes carry, two class names, and the shape of element ids. A
// name that disagrees is not an error anywhere: it is a `closest` that returns
// null, so a selection inside a page stops producing an "Explain" pill, or a
// canvas stops finding the host it was registered against, and the only symptom
// is a reader that quietly does nothing.
//
// The app's half is `src/dom_contract.rs`, which holds the names Rust uses as
// values. Two of them — `data-host-page` and `data-ai-popover` — cannot live
// there, because a Leptos view takes an attribute's name from the markup and
// only its value from an expression, so the hosts write them as literals; the
// check below reads those literals out of the Rust source instead.
//
// `tools/check-dom-contract.ts` fails CI when the halves disagree. It also
// reads the id shapes out of the Rust builders' `format!` strings rather than
// trusting a copy of them, so a rename on either side is caught by the other.
// Nothing outside this file should spell any of these names.

// --- attributes every reader host carries ---------------------------------

/** Names the format family that painted a host: `pdf` or `reflow`. */
export const HOST_ATTR = "data-reader-host";

/** The 1-based page a host is showing. The id shape is only the fallback. */
export const HOST_PAGE_ATTR = "data-host-page";

/** On a rendered block of a reflowable document: its index in document order. */
export const BLOCK_INDEX_ATTR = "data-block-index";

/** On the AI pill's root: a press here is not a click that clears a selection. */
export const AI_POPOVER_ATTR = "data-ai-popover";

// --- values `HOST_ATTR` carries -------------------------------------------

/**
 * The engine only ever branches on `reflow`: a PDF host is the path it has
 * always taken, and the app decides what to do with a `host` value it does not
 * recognise. `HOST_PDF` is therefore the app's to declare (`src/dom_contract.rs`),
 * and the check forbids spelling either value as a literal all the same.
 */
export const HOST_REFLOW = "reflow";

// --- class names both sides look up ---------------------------------------

/** The text layer inside a PDF host: the app builds it, the engine fills it. */
export const TEXT_LAYER_CLASS = "textLayer";

/** The still-bitmap overlay a zoom stretches while a re-render is on its way. */
export const PAGE_SNAPSHOT_CLASS = "page-snapshot";

// --- element id shapes ----------------------------------------------------
//
// `components/viewer/page_host.rs` builds these. Three of the four prefixes
// number their pages from 1; the continuous strip indexes its window from 0,
// and so do its ids, which is the one asymmetry a caller has to know.

export const ID_PREFIX_SINGLE = "sp";
export const ID_PREFIX_SPREAD = "dp";
export const ID_PREFIX_HSCROLL = "hp";
export const ID_PREFIX_STREAM = "cont";

/** Suffix of a page host's id. */
export const HOST_ID_SUFFIX = "-pg";
/** Suffix of the canvas inside a host: the host id with this instead. */
export const CANVAS_ID_SUFFIX = "-cv";
/** Suffix of the wrapper row around one page of a vertical strip. */
export const STREAM_WRAP_SUFFIX = "-wrap";

// --- derived selectors ----------------------------------------------------

export const HOST_SELECTOR = `[${HOST_ATTR}]`;
export const BLOCK_ROW_SELECTOR = `[${BLOCK_INDEX_ATTR}]`;
export const AI_POPOVER_SELECTOR = `[${AI_POPOVER_ATTR}]`;
export const TEXT_LAYER_SELECTOR = `.${TEXT_LAYER_CLASS}`;
export const PAGE_SNAPSHOT_SELECTOR = `.${PAGE_SNAPSHOT_CLASS}`;
/** A selection in the gap between two pages of the strip lands on a wrapper. */
export const STREAM_WRAP_SELECTOR = `[id^='${ID_PREFIX_STREAM}-'][id$='${STREAM_WRAP_SUFFIX}']`;

// --- id parsing -----------------------------------------------------------
//
// Compiled once, not per call: `pageFromCanvasId` runs on the fallback path of
// every unmounted-id resolution, which on a fast scroll is per row.

const PREFIX_GROUP = [ID_PREFIX_SINGLE, ID_PREFIX_SPREAD, ID_PREFIX_HSCROLL, ID_PREFIX_STREAM].join(
  "|"
);
const HOST_PAGE_RE = new RegExp(`^(${PREFIX_GROUP})-(\\d+)${HOST_ID_SUFFIX}$`);
const CANVAS_PAGE_RE = new RegExp(`^(${PREFIX_GROUP})-(\\d+)${CANVAS_ID_SUFFIX}$`);
const STREAM_WRAP_RE = new RegExp(`^${ID_PREFIX_STREAM}-(\\d+)${STREAM_WRAP_SUFFIX}$`);

/** 1-based page from a `^(prefix)-(n)$` match; the strip's ids are 0-based. */
function pageOf(prefix: string | undefined, raw: string | undefined): number | null {
  if (!raw) return null;
  const n = parseInt(raw, 10);
  if (!Number.isFinite(n) || n < 0) return null;
  return prefix === ID_PREFIX_STREAM ? n + 1 : n;
}

/** The 1-based page a host id names, or null when the id is not a page host's. */
export function pageFromHostId(id: string): number | null {
  const m = HOST_PAGE_RE.exec(id);
  return m ? pageOf(m[1], m[2]) : null;
}

/** The 1-based page a canvas id names, or null when the id is not a canvas's. */
export function pageFromCanvasId(id: string): number | null {
  const m = CANVAS_PAGE_RE.exec(id);
  return m ? pageOf(m[1], m[2]) : null;
}

/** The 1-based page a strip wrapper's id names, or null for any other id. */
export function pageFromWrapId(id: string): number | null {
  const m = STREAM_WRAP_RE.exec(id);
  return m ? pageOf(ID_PREFIX_STREAM, m[1]) : null;
}

/** The host a canvas id belongs to: the same id with the host suffix. */
export function hostIdFromCanvasId(canvasId: string): string {
  if (!canvasId.endsWith(CANVAS_ID_SUFFIX)) return canvasId;
  return canvasId.slice(0, -CANVAS_ID_SUFFIX.length) + HOST_ID_SUFFIX;
}

File: src/dom_contract.rs
//! The DOM vocabulary the app shares with the imperative engine.
//!
//! The engine under `public/engine/` paints PDF pages into hosts the app
//! builds, and the two sides meet only in the DOM: attribute names, the values
//! those attributes carry, one class name, and the shape of element ids.
//! Neither compiler can see the other, so a rename here is not an error there —
//! it is a query that quietly returns nothing. A host that stops advertising
//! itself is a selection that never becomes an "Info" pill; an id that changes
//! shape is a canvas that never finds the page it belongs to.
//!
//! What lives here is the half Rust uses as a *value*: in a `closest` selector,
//! in a `get_attribute`, in a comparison. Two more names cross the boundary and
//! cannot live here, because Leptos takes an attribute's NAME from the markup
//! and only its value from an expression:
//!
//! * `data-host-page` — the 1-based page a host paints, written by
//!   `crate::components::formats::pdf::canvas`,
//!   `crate::components::formats::reflow::page` and
//!   `crate::components::formats::reflow::stream`. It is the first thing the
//!   engine's selection tracker asks a host for; the id shape is only the
//!   fallback.
//! * `data-ai-popover` — marks the AI menu's root so that a press inside it is
//!   not read as a click that clears the selection
//!   (`crate::components::ai::selection_pill`).
//!
//! Element ids are built, not spelled: `crate::components::viewer::page_host`
//! owns [`crate::components::viewer::page_host::host_id_for_mode`] (the
//! `sp-`/`dp-`/`hp-`/`cont-` hosts and their `-cv` canvases) and
//! [`crate::components::viewer::page_host::block_row_id`], and
//! `crate::components::formats::pdf::strip` owns the `-wrap` rows around a
//! vertical strip's pages. Those shapes are part of this contract even though
//! the strings are not here — putting them here would mean writing them twice
//! in Rust as well.
//!
//! The engine's half of the table is `public/engine/dom-contract.ts`, and
//! `scripts/check-dom-contract.ts` fails CI when the two halves disagree. It
//! reads the id shapes out of the builders' `format!` strings rather than
//! trusting a second copy of them, so a rename in either language is caught by
//! the other.

/// The attribute every reader page host carries, naming the format family that
/// painted it. The engine's selection tracker and the app's own captures both
/// find their host through it, so a new format adds one attribute and joins.
pub const HOST_ATTR: &str = "data-reader-host";

/// The [`HOST_ATTR`] value a reflowable page or stream block carries.
pub const HOST_REFLOW: &str = "reflow";

/// The [`HOST_ATTR`] value a PDF page carries.
pub const HOST_PDF: &str = "pdf";

/// On a rendered block: which block of the document it is, in document order.
///
/// The identity half of the two handles a reflowable mark has on the DOM — the
/// engine's selection tracker walks up to it with `closest`, and a capture reads
/// the block number off it — and it is what makes the paginated modes and the
/// continuous stream resolve identically. The lookup half is the element id,
/// [`crate::components::viewer::page_host::block_row_id`], which the projection
/// resolves per mark per refresh.
pub const BLOCK_INDEX_ATTR: &str = "data-block-index";

/// The class of the still-bitmap overlay a zoom stretches while a re-render is
/// on its way.
///
/// `crate::components::formats::pdf::canvas_host` creates and removes it; the
/// engine's teardown clears any that outlived a document, because a snapshot
/// left in a recycled host would show the previous page's pixels.
pub const PAGE_SNAPSHOT_CLASS: &str = "page-snapshot";

/// The class of the text layer inside a PDF host.
///
/// The app builds the empty div (`crate::components::formats::pdf::canvas`); the
/// engine fills it with positioned spans, replaces it on a zoom, and queries it
/// when a selection has to know which layer of the document it is in. Both sides
/// look it up by this name.
pub const TEXT_LAYER_CLASS: &str = "textLayer";
